feat: Additional Features (#1345)
discordjs/guide: guide/additional-features/reloading-commands.md

# Reloading Commands

When writing your commands, you may find it tedious to restart your bot every time for testing the smallest changes. With a command handler, you can eliminate this issue and reload your commands while your bot is running.

This section assumes you followed the [Command Handling](/guide/creating-your-bot/command-handling.md) part.

::: warning
ESM does not support require and clearing import cache. You can use [hot-esm](https://www.npmjs.com/package/hot-esm) to import files without cache.
:::

```js {1}
const { SlashCommandBuilder } = require('discord.js');

module.exports = {
	data: new SlashCommandBuilder()

> 		.setName('reload')

Reviewer: i dont know if leaving it to use for everyone is a good idea
some warning etc should be here

# Reloading Commands

When writing your commands, you may find it tedious to restart your bot every time for testing the smallest changes. With a command handler, you can eliminate this issue and reload your commands while your bot is running.

::: warning
ESM does not support require and clearing import cache. You can use [hot-esm](https://www.npmjs.com/package/hot-esm) to import files without cache.
:::

This section assumes you followed the [Command Handling](/guide/creating-your-bot/command-handling.md) part.

::: warning

> The reload is not a command that should be used for every users. You should deploy it as a guild command in a private guild.

Reviewer 2: ```suggestion
The reload is not a command that should be generally used by every user. You should deploy it as a guild command in a private guild.
```

:::

> ```js {1}

Reviewer: ```suggestion
```js
```
pointing at imporing builder?

const { SlashCommandBuilder } = require('discord.js');

module.exports = {
	data: new SlashCommandBuilder()
		.setName('reload')
		.setDescription('Reloads a command.')
		.addStringOption(option =>
			option.setName('command')
				.setDescription('The command to reload.')
				.setRequired(true)),
	async execute(interaction) {
		// ...
	},
};
```

First off, you need to check if the command you want to reload exists. You can do this check similarly to getting a command.

> ```js {4-6,8-10}

Reviewer: ```suggestion
```js {4-5,8}
```

module.exports = {
	// ...
	async execute(interaction) {
		const commandName = interaction.options.getString('command', true).toLowerCase();
		const command = interaction.client.commands.get(commandName);

		if (!command) {
			return interaction.reply(`There is no command with name \`${commandName}\`!`);
		}
	},
};
```

To build the correct file path, you will need the file name. You can use `command.data.name` for doing that.

In theory, all there is to do is delete the previous command from `client.commands` and require the file again. In practice, you cannot do this easily as `require()` caches the file. If you were to require it again, you would load the previously cached file without any changes. You first need to delete the file from the `require.cache`, and only then should you require and set the command file to `client.commands`:

> ```js {3,5-12}

Reviewer: ```suggestion
```js {1,4-6}
```
